api/v1beta1/types_template.go

```diff
@@ -107,6 +109,8 @@ func (s SubnetTemplateSpec) IsNatGatewayEnabled() bool {
 }
 
 // SubnetTemplatesSpec specifies a list of subnet templates.
+// +listType=map
+// +listMapKey=role
```

Comment:
@CecileRobertMichon do you have any thoughts on this one? If `role` isn't sufficiently unique we'll probably have to define a union of properties for unique identification

Reply:
SubnetTemplate spec is only a subset of the SubnetSpec fields that are shared across clusters and was added for ClusterClass in order to be used in AzureClusterTemplate only. AFAIK there is no unique field combination in `SubnetTemplatesSpec` as-is... The uniqueness will only come from `name`, which will be generated to be a combination of cluster name + role + unique index.

TBH, I'm not too sure why `name` wasn't included in the template since you could have repeated subnet names that are the same across clusters, I think? Maybe @ykakarap remembers more than I do since he authored https://github.com/kubernetes-sigs/cluster-api-provider-azure/commit/99a589cef3e119ff7cbf890e9514da5309965dba

Reply:
For reference: `name` uniqueness validation is here: https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/b59c9198f45c3cb13a0b1209cb8167ebcc4156ac/api/v1beta1/azurecluster_validation.go#L180

Reply:
> TBH, I'm not too sure why name wasn't included in the template since you could have repeated subnet names that are the same across clusters, I think? Maybe @ykakarap remembers more than I do since he authored https://github.com/kubernetes-sigs/cluster-api-provider-azure/commit/99a589cef3e119ff7cbf890e9514da5309965dba

Yup. If we could introduce name here as well and make it the map key, this would solve all our problems.

@ykakarap do you know if there was a strong reason not to add it here?

Reply:
> The uniqueness will only come from name, which will be generated to be a combination of cluster name + role + unique index.

@CecileRobertMichon  That's how it's done in the AzureCluster webhook, right? Would it be a problem if the names are hard-coded to unique values in the AzureClusterTemplate and then used for multiple clusters? (aka does the name have to be unique across clusters?)

Reply:
>  this would solve all our problems.

I like the sound of that

> Would it be a problem if the names are hard-coded to unique values in the AzureClusterTemplate and then used for multiple clusters? (aka does the name have to be unique across clusters?)

no it shouldn't be, subnet name just needs to be unique per cluster (ie. two subnets in the same cluster can't have the same name), maybe that's where the confusion came from

Reply:
@CecileRobertMichon Found this comment: https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/1501/files#r673287133

Do you know why? :)

Reply:
So in Azure, there are naming scopes based on each resource type: https://learn.microsoft.com/en-us/azure/cloud-adoption-framework/ready/azure-best-practices/resource-naming#example-names-networking

For subnets, the scope is vnet. Meaning two subnets can share the same name as long as they are in different vnets.

For vnets, the scope is resource group. Meaning two vnets can share the same name as long as they are in different resource groups.

For resource groups, the scope is subscription. Meaning two resource groups can share the same name as long as they are in different subscriptions.

So in theory, a cluster class template could have hard-coded unique subnet names as long as a different vnet and/or resource group and/or subscription is used for each cluster. I think each cluster in a clusterclass template should definitely have its own resource group, because chances of using a different subscription for each cluster in practice are pretty low. And if that's the case (one resource group == one cluster), then we can safely reuse subnet names across clusters.

Does that make sense?

Reply:
Sounds good to me thx

Reply:
>@ykakarap do you know if there was a strong reason not to add it here?

Did that because the last time we had a discussion about thing when introducing the AzureClusterTemplate it was told that the names had to be unique (thanks for the link to the old comment @sbueringer). That was the only reason. I do not recall any other reason for doing that. 

If the names can be duplicate among multiple clusters then I don't see a problem with making it part of `SubnetTemplatesSpec`.

**Note:** Once we move `Name` into `SubnetTemplatesSpec` we will need to introduce validation in `[validateSubnetTemplates](https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/ecf9115bb4230527f8a93098b8590f5e5d15443e/api/v1beta1/azureclustertemplate_validation.go#L77-L78)` to make sure that there are no duplicate name within the same AzureClusterTemplate.

**Note:** We will also need to introduce defaulting for this field in AzureClusterTemplate. Otherwise it could lead to infinite reconcile (?).

Reply:
@ykakarap Thx for the infos!

> Note: We will also need to introduce defaulting for this field in AzureClusterTemplate. Otherwise it could lead to infinite reconcile (?).

I'm not sure we have to introduce defaulting, but I'm definitely not against it. By introducing `name` as a mandatory field it always has to be set. So I'm not sure how this would lead to infinite reconciles. I would probably lean towards not having defaulting so that the subnet name really has to be defined in the AzureClusterTemplate which should make it easier to match the config in AzureClusterTemplate to the resulting AzureCluster (as the names would have to exist in the local AzureClusterTemplate YAML and in the resulting AzureClusters).

If I understand the current state in AzureCluster correctly the field is mandatory there but users don't have to set it on client-side as the webhook auto-generates the name if they don't exist

Reply:
Oh you are right! Since it is mandatory field we don't have do defaulting. Just validation should be enough.